### DISCUSSION TITLE
[vite-node] Is it for running code that you would build with Vite and run in production?

### DISCUSSION THREAD

**Discussion by silvenon** [vitest-dev/vitest#5970] [Q&A] [answered]
I'm aware of Vite's Environment API proposal, but since this is a working prior art, I'm trying to figure out what vite-node is useful for exactly.

I have an Express server code that's written TypeScript, so far I had been running it with [tsx](https://github.com/privatenumber/tsx), but I would like to move to building that server code with Vite. However, in that case running it with tsx will no longer cut it because Vite stuff won't work during development in the server code.

It was my understanding that vite-node is good for this use case, but I was hoping to run into problems that I would have run into upon building my server code and running with Node.js, but that wasn't the case. I was able to successfully run with vite-node, and fail to run the built server code with Node.js because of ESM/CJS differences etc.

vite-node seems to take some stuff from my Vite config, but probably not `build` part of the config, right? Is there a way to use vite-node to solve my problem of wanting to see basic runtime errors right away? vite-node solves `__dirname` and `__filename`, but I don't want that problem solved because it won't work in the built code, so maye it isn't the tool for my use case after all?

Although it's pretty damn handy ❤️

**Comment by sheremet-va** [comment, marked as answer]
I wouldn't recommend using vite-node for production since it has a lot of overhead of using `vm` and spinning up a Vite server which also increases the memory usage that's not related to your application. This is also why `build` part is ignored - vite-node uses dev transforms to run your code.

For dev use case, we are not planning to remove `__filename` and `__dirname` shims since it's a feature. What you can do is create your own small wrapper around `vite.ssrLoadModule` since it tries to be as compatible with Node.js as possible (no shims, no CJS/ESM proxies):

```ts
import { createServer } from 'vite'
const server = createServer()
await server.ssrLoadModule(process.argv[0])
```

**Reply by silvenon** [reply]
Yep, I want those shims in Vitest, so I wasn't suggesting to remove them.

The approach you recommended seems like the right one, I'll take some time to hash out the details of figuring out watch mode, HMR etc. Thankfully, I have vite-node's source code for crib notes.

Thanks for pointing me in the direction that might work for me, I got really tangled up.
